File: go-daemon/oshooks.go
package main

import (
	"fmt"
	"os/exec"
	"runtime"
)

// RevealInOS highlights path in the native file browser: Finder on macOS,
// File Explorer on Windows. Called when the UI's "System Hook Actions"
// button is pressed on a search result.
// Start(), not Run() — Run() blocks the caller until the child process
// fully exits, which only matters if something needs its exit code. It
// doesn't here: revealing a file in Finder/Explorer is fire-and-forget,
// and waiting on it needlessly stalls whatever called this (see
// rust-engine/src/oshooks.rs's mirrored fix — that one was live-wired to
// the HTTP handler and was actually stalling the whole engine).
func RevealInOS(path string) error {
	switch runtime.GOOS {
	case "darwin":
		return exec.Command("open", "-R", path).Start()
	case "windows":
		// explorer.exe returns a non-zero exit code on success in some
		// versions; ignore the error value deliberately and only surface
		// a real failure to start the process.
		return exec.Command("explorer.exe", "/select,"+path).Start()
	default:
		return fmt.Errorf("RevealInOS: unsupported platform %s", runtime.GOOS)
	}
}

File: rust-engine/src/oshooks.rs
//! Native "reveal in file browser" hook, mirrored from go-daemon/oshooks.go.
//! Lives here too because the browser-based frontend's HTTP `/api/reveal`
//! call is served directly by the engine, not proxied through the daemon —
//! the daemon's copy remains for parity if a future native-client path
//! calls IndexService/SearchService directly instead of the HTTP shim.

use anyhow::Result;
#[cfg(not(any(target_os = "macos", target_os = "windows")))]
use anyhow::bail;
use std::process::Command;

pub fn reveal(path: &str) -> Result<()> {
    #[cfg(target_os = "macos")]
    {
        // .spawn(), not .status() — status() blocks the calling thread
        // until `open` fully exits, and this runs straight inside an async
        // HTTP handler with no spawn_blocking wrapper, so it was stalling
        // the engine's whole tokio runtime (search, status, everything)
        // for however long Finder took to actually respond, not just the
        // near-instant fork+exec this only needs. We don't care about the
        // exit code — reveal-in-Finder is inherently fire-and-forget.
        Command::new("open").arg("-R").arg(path).spawn()?;
        return Ok(());
    }

    #[cfg(target_os = "windows")]
    {
        Command::new("explorer.exe").arg(format!("/select,{path}")).spawn()?;
        return Ok(());
    }

    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
    {
        bail!("reveal-in-file-browser is only supported on macOS and Windows");
    }
}
